### PR TITLE
Open id connect discovery

### PR DESCRIPTION
**Pull request by ChamaraPhilipsuom**
Implementation of Open ID Connect Discovery. Code reiview.

### CODE REVIEW

File: components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/webfinger/impl/WebFingerJSOnResponseBuilder.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.carbon.identity.oauth.endpoint.webfinger.impl;
+
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.codehaus.jettison.json.JSONException;
```
Comment (hasinthaindrajee) [inline components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/webfinger/impl/WebFingerJSOnResponseBuilder.java]: avoid star imports

Comment (ChamaraPhilipsuom) [inline components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/webfinger/impl/WebFingerJSOnResponseBuilder.java]: Edited the file removing star imports.

File: components/oauth/org.wso2.carbon.identity.discovery/pom.xml [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+                            org.wso2.carbon.base; version="${carbon.base.imp.pkg.version.range}",
+                            org.wso2.carbon.identity.oauth2.*;
+                            version="${carbon.identity.package.import.version.range}",
+                        </Import-Package>
+                        <Export-Package>
+                            !org.wso2.carbon.identity.discovery.internal,
+                            org.wso2.carbon.identity.discovery.*;
+                            version="${carbon.identity.package.export.version}"
+                        </Export-Package>
+                        <DynamicImport-Package>*</DynamicImport-Package>
```
Comment (hasinthaindrajee) [inline components/oauth/org.wso2.carbon.identity.discovery/pom.xml]: Any specific reasons to use dynamic imports  ?

File: components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/DefaultOIDCProcessor.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+    public int handleError(OIDCDiscoveryEndPointException error) {
+        if (log.isDebugEnabled()) {
+            log.debug(error);
+        }
+        String errorCode = error.getErrorCode();
+        if (errorCode.equals(OIDCDiscoveryEndPointException.ERROR_CODE_NO_OPENID_PROVIDER_FOUND)) {
+
+        } else if (errorCode.equals(OIDCDiscoveryEndPointException.ERROR_CODE_INVALID_REQUEST)) {
+
```
Comment (hasinthaindrajee) [inline components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/DefaultOIDCProcessor.java]: Reason for ignoring these cases ?

File: components/oauth/org.wso2.carbon.identity.discovery/pom.xml [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.apache.amber.wso2</groupId>
+            <artifactId>amber</artifactId>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.wso2.carbon.identity</groupId>
+            <artifactId>org.wso2.carbon.identity.oauth</artifactId>
+            <scope>provided</scope>
```
Comment (pulasthi7) [inline components/oauth/org.wso2.carbon.identity.discovery/pom.xml]: do we need to add the scope?

File: components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/DefaultOIDCProcessor.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        } else if (errorCode.equals(OIDCDiscoveryEndPointException.ERROR_CODE_INVALID_TENANT)) {
+
+        } else if (errorCode.equals(OIDCDiscoveryEndPointException.ERROR_CODE_JSON_EXCEPTION)) {
+
+        } else {
+            return HttpServletResponse.SC_INTERNAL_SERVER_ERROR;
+        }
```
Comment (pulasthi7) [inline components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/DefaultOIDCProcessor.java]: what is the use of these if conditions? their bodies are empty and the method will always return HttpServletResponse.SC_INTERNAL_SERVER_ERROR ?? If so can't we get rid of L57-L68 ?

File: components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/DiscoveryConstants.java [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+    value is true. If omitted, the default value is false.
+op_policy_uri
+    OPTIONAL. URL that the OpenID Provider provides to the person registering the Client to read about
+    the OP's requirements on how the Relying Party can use the data provided by the OP. The registration
+    process SHOULD display this URL to the person registering the Client if it is given.
+op_tos_uri
+    OPTIONAL. URL that the OpenID Provider provides to the person registering the Client to read about
+    OpenID Provider's terms of service. The registration process SHOULD display this URL to the person
+    registering the Client if it is given.
+*/
```
Comment (pulasthi7) [inline components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/DiscoveryConstants.java]: Can't we have these as doc comments for each constant below separately?

File: components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/OIDCProcessor.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+ * under the License.
+ */
+package org.wso2.carbon.identity.discovery;
+
+import org.wso2.carbon.base.ServerConfigurationException;
+
+import javax.servlet.http.HttpServletRequest;
+
+public interface OIDCProcessor {
+    public OIDProviderConfigResponse getResponse(HttpServletRequest request, String tenant) throws
```
Comment (pulasthi7) [inline components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/OIDCProcessor.java]: better to make the variable name as tenantDomain

File: components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/OIDProviderRequest.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.carbon.identity.discovery;
+
+
+public class OIDProviderRequest {
+
+    private String uri;
+    private String tenant;
```
Comment (pulasthi7) [inline components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/OIDProviderRequest.java]: rename variable to tenantDomain

File: components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/builders/DefaultOIDProviderRequestBuilder.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+
+public class DefaultOIDProviderRequestBuilder implements OIDProviderRequestBuilder {
+
+    @Override
+    public OIDProviderRequest buildRequest(HttpServletRequest request, String tenant) throws
+            OIDCDiscoveryEndPointException {
+        OIDProviderRequest requestObject = new OIDProviderRequest();
+        requestObject.setUri(request.getRequestURI());
+        if (tenant != null && !tenant.isEmpty()) {
```
Comment (pulasthi7) [inline components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/builders/DefaultOIDProviderRequestBuilder.java]: use StringUtils

File: components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/oidcdiscovery/OIDCDiscoveryEndpoint.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        OIDCProcessor processor = EndpointUtil.getOIDCService();
+        try {
+            OIDProviderResponseBuilder responseBuilder = new OIDProviderJSONResponseBuilder();
+            response = responseBuilder.getOIDProviderConfigString(processor.getResponse(request, tenant));
+        } catch (OIDCDiscoveryEndPointException e) {
+            Response.ResponseBuilder errorResponse = Response.status(processor.handleError(e));
+            return errorResponse.entity(e.getMessage()).build();
+        } catch (ServerConfigurationException e) {
+            Response.ResponseBuilder errorResponse = Response.status(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+            return errorResponse.entity("Error in reading configuration.").build();
```
Comment (pulasthi7) [inline components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/oidcdiscovery/OIDCDiscoveryEndpoint.java]: log 'e'

File: components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/webfinger/WebFingerEndPoint.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+        String response;
+        try {
+            ResponseBuilder responseBuilder = new JSONResponseBuilder();
+            response = responseBuilder.getOIDProviderIssuerString(processor.getResponse(request));
+        } catch (WebFingerEndpointException e) {
+            Response.ResponseBuilder errorResponse = Response.status(processor.handleError(e));
+            return errorResponse.entity(e.getErrorMessage()).build();
+        } catch (ServerConfigurationException e) {
+            Response.ResponseBuilder errorResponse = Response.status(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+            return errorResponse.entity("Error in reading configuration.").build();
```
Comment (pulasthi7) [inline components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/webfinger/WebFingerEndPoint.java]: log 'e'

File: components/oauth/org.wso2.carbon.identity.webfinger/pom.xml [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.apache.amber.wso2</groupId>
+            <artifactId>amber</artifactId>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.wso2.carbon.identity</groupId>
+            <artifactId>org.wso2.carbon.identity.oauth</artifactId>
+            <scope>provided</scope>
```
Comment (pulasthi7) [inline components/oauth/org.wso2.carbon.identity.webfinger/pom.xml]: do we need scope here?

File: components/oauth/org.wso2.carbon.identity.webfinger/src/main/java/org/wso2/carbon/identity/webfinger/DefaultWebFingerProcessor.java [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+            log.debug(error);
+        }
+        String errorCode = error.getErrorCode();
+        if (errorCode.equals(WebFingerConstants.ERROR_CODE_INVALID_REQUEST)) {
+            return HttpServletResponse.SC_BAD_REQUEST;
+        } else if (errorCode.equals(WebFingerConstants.ERROR_CODE_INVALID_RESOURCE)) {
+            return HttpServletResponse.SC_NOT_FOUND;
+        } else if (errorCode.equals(WebFingerConstants.ERROR_CODE_JSON_EXCEPTION)) {
+            return HttpServletResponse.SC_UNSUPPORTED_MEDIA_TYPE;
+        } else if (errorCode.equals(WebFingerConstants.ERROR_CODE_NO_WEBFINGER_CONFIG)) {
```
Comment (pulasthi7) [inline components/oauth/org.wso2.carbon.identity.webfinger/src/main/java/org/wso2/carbon/identity/webfinger/DefaultWebFingerProcessor.java]: use "constant.equals(variable)" pattern

File: components/oauth/org.wso2.carbon.identity.webfinger/src/main/java/org/wso2/carbon/identity/webfinger/URLNormalizer.java [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+import org.wso2.carbon.user.api.UserStoreException;
+
+import java.net.URI;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+
+public class URLNormalizer {
+
+
```
Comment (pulasthi7) [inline components/oauth/org.wso2.carbon.identity.webfinger/src/main/java/org/wso2/carbon/identity/webfinger/URLNormalizer.java]: remove extra new lines

File: components/oauth/org.wso2.carbon.identity.webfinger/src/main/java/org/wso2/carbon/identity/webfinger/URLNormalizer.java [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+    public static void validateTenant(String userInfo) throws WebFingerEndpointException {
+        try {
+            int tenantId = WebFingerServiceComponentHolder.getRealmService().getTenantManager().getTenantId(userInfo);
+            if (tenantId < 0 && tenantId!= -1234) {
+                throw new WebFingerEndpointException(WebFingerConstants.ERROR_CODE_INVALID_TENANT, "The tenant domain" +
+                        " is not valid.");
+            }
+        } catch (UserStoreException e) {
+            throw new WebFingerEndpointException(WebFingerConstants.ERROR_CODE_INVALID_TENANT, e.getMessage());
```
Comment (pulasthi7) [inline components/oauth/org.wso2.carbon.identity.webfinger/src/main/java/org/wso2/carbon/identity/webfinger/URLNormalizer.java]: 'e' is not logged or thrown

File: components/oauth/org.wso2.carbon.identity.webfinger/src/main/java/org/wso2/carbon/identity/webfinger/URLNormalizer.java [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+            return request;
+        }
+
+
+    }
+
+    public static void validateTenant(String userInfo) throws WebFingerEndpointException {
+        try {
+            int tenantId = WebFingerServiceComponentHolder.getRealmService().getTenantManager().getTenantId(userInfo);
+            if (tenantId < 0 && tenantId!= -1234) {
```
Comment (pulasthi7) [inline components/oauth/org.wso2.carbon.identity.webfinger/src/main/java/org/wso2/carbon/identity/webfinger/URLNormalizer.java]: use defined constants in MultitenantConstatnts

File: components/oauth/org.wso2.carbon.identity.webfinger/src/main/java/org/wso2/carbon/identity/webfinger/builders/WebFingerResponseBuilder.java [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+
+    public WebFingerResponse buildWebFingerResponse(WebFingerRequest request) throws WebFingerEndpointException,
+            ServerConfigurationException {
+        WebFingerResponse response = this.getLinkFromIdentityConfig(request.getRel(), request.getResource());
+        if (response == null) {
+            try {
+                response = this.getIssuerFromServerURL(request.getResource());
+            } catch (IdentityException e) {
+                throw new WebFingerEndpointException(WebFingerConstants.ERROR_CODE_NO_WEBFINGER_CONFIG, "Error in " +
+                        "getting server url.\n" + e.getMessage());
```
Comment (pulasthi7) [inline components/oauth/org.wso2.carbon.identity.webfinger/src/main/java/org/wso2/carbon/identity/webfinger/builders/WebFingerResponseBuilder.java]: 'e' is not logged or thrown

File: components/oauth/org.wso2.carbon.identity.webfinger/src/main/java/org/wso2/carbon/identity/webfinger/builders/WebFingerResponseBuilder.java [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+
+    /**
+     * This is not generic. This is intended to get the web finger issuer only.
+     *
+     * @return WebFingerResponse with http://openid.net/specs/connect/1.0/issuer, ServerUrl added ti the WebLink list.
+     * @resource resource parameter in the web finger request
+     */
+    private WebFingerResponse getIssuerFromServerURL(String resource) throws IdentityException {
+        String issuer = IdentityUtil.getServerURL("");
+        if (issuer == null || issuer.isEmpty()) {
```
Comment (pulasthi7) [inline components/oauth/org.wso2.carbon.identity.webfinger/src/main/java/org/wso2/carbon/identity/webfinger/builders/WebFingerResponseBuilder.java]: use StringUtils

File: components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/DefaultOIDCProcessor.java (hunk 3 above)
Comment (ChamaraPhilipsuom) [inline components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/DefaultOIDCProcessor.java]: The error codes are not specified in the spec. Should return the proper error code. This was left as a future work. I have added logs at https://github.com/wso2/carbon-identity/pull/1694. Shall I add @TODO

File: components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/DefaultOIDCProcessor.java (hunk 5 above)
Comment (ChamaraPhilipsuom) [inline components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/DefaultOIDCProcessor.java]: The error codes are not specified in the spec. Should return the proper error code. This was left as a future work. I have added logs at #1694. Shall I add @TODO?

File: components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/DiscoveryConstants.java (hunk 6 above)
Comment (ChamaraPhilipsuom) [inline components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/DiscoveryConstants.java]: Fixed at #1694

File: components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/OIDCProcessor.java (hunk 7 above)
Comment (ChamaraPhilipsuom) [inline components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/OIDCProcessor.java]: Fixed at #1694

File: components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/OIDProviderRequest.java (hunk 8 above)
Comment (ChamaraPhilipsuom) [inline components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/OIDProviderRequest.java]: Fixed at #1694

File: components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/builders/DefaultOIDProviderRequestBuilder.java (hunk 9 above)
Comment (ChamaraPhilipsuom) [inline components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/builders/DefaultOIDProviderRequestBuilder.java]: Fixed at #1694

File: components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/oidcdiscovery/OIDCDiscoveryEndpoint.java (hunk 10 above)
Comment (ChamaraPhilipsuom) [inline components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/oidcdiscovery/OIDCDiscoveryEndpoint.java]: OIDCDiscoveryEndpointException is logged at OIDCProcessor.handleError().
ServerConfigurationException is loged at #1694

File: components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/webfinger/WebFingerEndPoint.java (hunk 11 above)
Comment (ChamaraPhilipsuom) [inline components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/webfinger/WebFingerEndPoint.java]: WebFingerEndpointException is logged at DefaultWebFingerProcessor.handleError().
ServerConfigurationException is loged at #1694

File: components/oauth/org.wso2.carbon.identity.webfinger/src/main/java/org/wso2/carbon/identity/webfinger/DefaultWebFingerProcessor.java (hunk 13 above)
Comment (ChamaraPhilipsuom) [inline components/oauth/org.wso2.carbon.identity.webfinger/src/main/java/org/wso2/carbon/identity/webfinger/DefaultWebFingerProcessor.java]: Fixed at #1694

File: components/oauth/org.wso2.carbon.identity.webfinger/src/main/java/org/wso2/carbon/identity/webfinger/URLNormalizer.java (hunk 14 above)
Comment (ChamaraPhilipsuom) [inline components/oauth/org.wso2.carbon.identity.webfinger/src/main/java/org/wso2/carbon/identity/webfinger/URLNormalizer.java]: Fixed at #1694

File: components/oauth/org.wso2.carbon.identity.webfinger/src/main/java/org/wso2/carbon/identity/webfinger/URLNormalizer.java (hunk 15 above)
Comment (ChamaraPhilipsuom) [inline components/oauth/org.wso2.carbon.identity.webfinger/src/main/java/org/wso2/carbon/identity/webfinger/URLNormalizer.java]: Fixed at #1694

File: components/oauth/org.wso2.carbon.identity.webfinger/src/main/java/org/wso2/carbon/identity/webfinger/URLNormalizer.java (hunk 16 above)
Comment (ChamaraPhilipsuom) [inline components/oauth/org.wso2.carbon.identity.webfinger/src/main/java/org/wso2/carbon/identity/webfinger/URLNormalizer.java]: Fixed at #1694

File: components/oauth/org.wso2.carbon.identity.webfinger/src/main/java/org/wso2/carbon/identity/webfinger/builders/WebFingerResponseBuilder.java (hunk 17 above)
Comment (ChamaraPhilipsuom) [inline components/oauth/org.wso2.carbon.identity.webfinger/src/main/java/org/wso2/carbon/identity/webfinger/builders/WebFingerResponseBuilder.java]: Fixed at #1694

File: components/oauth/org.wso2.carbon.identity.webfinger/src/main/java/org/wso2/carbon/identity/webfinger/builders/WebFingerResponseBuilder.java (hunk 18 above)
Comment (ChamaraPhilipsuom) [inline components/oauth/org.wso2.carbon.identity.webfinger/src/main/java/org/wso2/carbon/identity/webfinger/builders/WebFingerResponseBuilder.java]: Fixed at #1694

File: components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/DefaultOIDCProcessor.java [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.base.ServerConfigurationException;
+import org.wso2.carbon.identity.discovery.builders.DefaultOIDProviderRequestBuilder;
+import org.wso2.carbon.identity.discovery.builders.OIDProviderRequestBuilder;
+import org.wso2.carbon.identity.discovery.builders.ProviderConfigBuilder;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class DefaultOIDCProcessor implements OIDCProcessor {
```
Comment (Kishanthan) [inline components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/DefaultOIDCProcessor.java]: Can we have class level javadoc on all the classes that are missing those? Please fix them throughout all your PRs.

File: components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/OIDProviderConfigResponse.java [hunk 20]
```diff
@@ -0,0 +1,401 @@
...
+
+    public String getIssuer() {
+        return issuer;
+    }
+
+    public void setIssuer(String issuer) {
+        this.issuer = issuer;
+    }
+
+    public String getAuthorization_endpoint() {
```
Comment (Kishanthan) [inline components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/OIDProviderConfigResponse.java]: What is the reason to use these type of method naming convention? Is it required at OID technical level?

File: components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/internal/OIDCDiscoveryServiceComponent.java [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+import org.wso2.carbon.identity.discovery.OIDCProcessor;
+
+/**
+ * @scr.component name="identity.discovery.component" immediate="true"
+ */
+public class OIDCDiscoveryServiceComponent {
+    private static Log log = LogFactory.getLog(OIDCDiscoveryServiceComponent.class);
+    private static BundleContext bundleContext = null;
+
+    public static BundleContext getBundleContext() {
```
Comment (Kishanthan) [inline components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/internal/OIDCDiscoveryServiceComponent.java]: This is a wrong way to get hold of bundleCtx instance. You should always use DataHolder pattern to store and retrieve object instances.

File: components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java [hunk 22]
```diff
@@ -57,6 +63,35 @@ private EndpointUtil() {
     }
 
     /**
+     * Returns the {@code DefaultWebFingerProcessor} instance
+     *
+     * @return
+     */
+    public static DefaultWebFingerProcessor getWebFingerService() {
```
Comment (Kishanthan) [inline components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java]: Why can't we use OSGi declarative service components to get the service references?

File: components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/OIDProviderConfigResponse.java (hunk 20 above)
Comment (ChamaraPhilipsuom) [inline components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/OIDProviderConfigResponse.java]: No. It was a mistake by using the parameter values as given in the spec. Fixed at #1694

File: components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/DefaultOIDCProcessor.java (hunk 19 above)
Comment (ChamaraPhilipsuom) [inline components/oauth/org.wso2.carbon.identity.discovery/src/main/java/org/wso2/carbon/identity/discovery/DefaultOIDCProcessor.java]: Fixed at #1694

**Pull request closed without merging**